index.js

```diff
@@ -0,0 +1,156 @@
+const mongo = require('mongodb');
+const async = require('async');
+const _ = require('lodash');
+const automap = require('automap');
+const defaults = {
+  mongo: {
+    host: 'mongodb://localhost:27017',
+    collectionName: 'hapi-activities'
+  },
+  timeout: 30 * 1000, // max time an action can take, default is 30 secs, set to false for infinity
+  interval: 5 * 60 * 1000, // 5 minutes
+  log: false
+};
+
+exports.register = (server, options, next) => {
+  const settings = _.defaults(options, defaults);
+  // connect to db:
+  mongo.connect(settings.mongo.host, (err, db) => {
+    if (err) {
+      return next(err);
+    }
+    // initialize the server object:
+    const collection = db.collection(settings.mongo.collectionName);
+
+    // update all activities:
+    const updateActivities = (allDone) => {
+      automap(
+        // fetch all 'waiting' activities and mark them as 'processing':
+        (done) => {
+          collection
+          .find({ status: 'waiting' })
+          .toArray((dbErr, results) => {
+            if (dbErr) {
+              server.log(['hapi-activities', 'error'], dbErr);
+            }
+            // can go back to sleep if nothing was found:
+            if (results.length === 0) {
+              return allDone();
+            }
+            // otherwise mark them all as 'processing' with one db call and proceed to process them:
+            const ids = _.reduce(results, (memo, result) => {
+              memo.push(result._id);
+              return memo;
+            }, []);
+            collection.update({ _id: { $in: ids } }, { $set: { status: 'processing' } }, { multi: true }, () => {
+              done(null, results);
+            });
+          });
+        },
+        // for each activity we just fetched and marked, do the following:
+        (activity) => {
+          return {
+            // log that it's underway:
+            logActivity: (done) => {
+              if (settings.log) {
+                server.log(['hapi-activities', 'starting-activity', 'debug'], { message: 'Processing underway for activity', data: activity });
+              }
+              done();
+            },
+            // execute the actions in parallel:
+            performActions: ['logActivity', (results, done) => {
+              // will launch the activity's actions in parallel:
+              const updatedActivity = {
+                results: []
+              };
+              async.each(settings.activities[activity.activityName], (action, eachDone) => {
```

Comment:
we need to make sure this still works if no activities are defined for that name